File: feature/movies/src/main/java/com/hari/tmdb/movie/item/CarouselGroup.java
package com.hari.tmdb.movie.item;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.xwray.groupie.Group;
import com.xwray.groupie.GroupAdapter;
import com.xwray.groupie.GroupDataObserver;
import com.xwray.groupie.Item;

/**
 * A group that contains a single carousel item and is empty when the carousel is empty
 **/
public class CarouselGroup implements Group {

    private boolean isEmpty;
    private GroupDataObserver groupDataObserver;
    private CarouselItem carouselItem;

    public CarouselGroup(RecyclerView.ItemDecoration itemDecoration, GroupAdapter adapter) {
        carouselItem = new CarouselItem(itemDecoration, adapter);
        isEmpty = adapter.getItemCount() == 0;
        RecyclerView.AdapterDataObserver adapterDataObserver = new RecyclerView.AdapterDataObserver() {

            @Override
            public void onItemRangeRemoved(int positionStart, int itemCount) {
                boolean empty = adapter.getItemCount() == 0;
                if (empty && !isEmpty) {
                    isEmpty = true;
                    groupDataObserver.onItemRemoved(carouselItem, 0);
                }
            }

            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                boolean empty = adapter.getItemCount() == 0;
                if (isEmpty && !empty) {
                    isEmpty = false;
                    groupDataObserver.onItemInserted(carouselItem, 0);
                }
            }
        };
        adapter.registerAdapterDataObserver(adapterDataObserver);
    }

    @Override
    public int getItemCount() {
        return isEmpty ? 0 : 1;
    }

    @NonNull
    @Override
    public Item getItem(int position) {
        if (position == 0 && !isEmpty) return carouselItem;
        else throw new IndexOutOfBoundsException();
    }

    @Override
    public int getPosition(@NonNull Item item) {
        return item == carouselItem && !isEmpty ? 0 : -1;
    }

    @Override
    public void registerGroupDataObserver(@NonNull GroupDataObserver groupDataObserver) {
        this.groupDataObserver = groupDataObserver;
    }

    @Override
    public void unregisterGroupDataObserver(@NonNull GroupDataObserver groupDataObserver) {
        this.groupDataObserver = null;
    }
}

File: corecomponent/androidcomponets/src/main/java/com/hari/tmdb/ui/item/CarouselGroup.kt
package com.hari.tmdb.ui.item

import androidx.recyclerview.widget.RecyclerView
import com.xwray.groupie.Group
import com.xwray.groupie.GroupAdapter
import com.xwray.groupie.GroupDataObserver
import com.xwray.groupie.Item

/**
 * A group that contains a single carousel item and is empty when the carousel is empty
 */
class CarouselGroup(
    itemDecoration: RecyclerView.ItemDecoration? = null,
    adapter: GroupAdapter<*>,
    layoutManager: RecyclerView.LayoutManager?
) : Group {
    private var isEmpty: Boolean
    private var groupDataObserver: GroupDataObserver? = null
    private val carouselItem: CarouselItem = CarouselItem(itemDecoration, adapter, layoutManager!!)

    override fun getItemCount(): Int {
        return if (isEmpty) 0 else 1
    }

    override fun getItem(position: Int): Item<*> {
        return if (position == 0 && !isEmpty) carouselItem else throw IndexOutOfBoundsException()
    }

    override fun getPosition(item: Item<*>): Int {
        return if (item === carouselItem && !isEmpty) 0 else -1
    }

    override fun registerGroupDataObserver(groupDataObserver: GroupDataObserver) {
        this.groupDataObserver = groupDataObserver
    }

    override fun unregisterGroupDataObserver(groupDataObserver: GroupDataObserver) {
        this.groupDataObserver = null
    }

    init {
        isEmpty = adapter.itemCount == 0
        val adapterDataObserver: RecyclerView.AdapterDataObserver =
            object : RecyclerView.AdapterDataObserver() {
                override fun onItemRangeRemoved(
                    positionStart: Int,
                    itemCount: Int
                ) {
                    val empty = adapter.itemCount == 0
                    if (empty && !isEmpty) {
                        isEmpty = true
                        groupDataObserver!!.onItemRemoved(carouselItem, 0)
                    }
                }

                override fun onItemRangeInserted(
                    positionStart: Int,
                    itemCount: Int
                ) {
                    val empty = adapter.itemCount == 0
                    if (isEmpty && !empty) {
                        isEmpty = false
                        groupDataObserver!!.onItemInserted(carouselItem, 0)
                    }
                }
            }
        adapter.registerAdapterDataObserver(adapterDataObserver)
    }
}